### PR TITLE
Add support for Kotlin `@Throws` in metadata

### PR DESCRIPTION
**Pull request by ZacSweers**
This adds support for detecting `@Throws` in metadata. Like some other jvm-specific things (certain modifiers, etc), these have to be inferred from the jvm code as the annotations are not present

### CODE REVIEW

File: kotlinpoet-metadata-specs/src/main/kotlin/com/squareup/kotlinpoet/metadata/specs/ElementHandler.kt [hunk 1]
```diff
@@ -129,6 +130,21 @@ interface ElementHandler {
...
   fun methodAnnotations(classJvmName: String, methodSignature: JvmMethodSignature): List<AnnotationSpec>
 
+  /**
+   * Looks up a given class method given its [JvmMethodSignature] and returns any thrown types
+   * found on it. Used for [Throws][@Throws]
+   *
+   * @param classJvmName The JVM name of the class (example: `"org/foo/bar/Baz$Nested"`).
+   * @param methodSignature The method with annotations to look up.
+   * @param isConstructor Indicates if [methodSignature] is a constructor.
+   * @return the set of found thrown types, or null.
```
Comment (Egorand) [inline kotlinpoet-metadata-specs/src/main/kotlin/com/squareup/kotlinpoet/metadata/specs/ElementHandler.kt]: nit: Any upside in returning null vs an empty set?

File: kotlinpoet-metadata-specs-tests/src/test/kotlin/com/squareup/kotlinpoet/metadata/specs/test/KotlinPoetMetadataSpecsTest.kt [hunk 2]
```diff
@@ -1192,6 +1192,47 @@ class KotlinPoetMetadataSpecsTest(
...
     }
   }
+
+  @Test
+  fun throws() {
+    val typeSpec = Throwing::class.toTypeSpecWithTestHandler()
+
+    //language=kotlin
+    assertThat(typeSpec.trimmedToString()).isEqualTo("""
+      class Throwing {
```
Comment (Egorand) [inline kotlinpoet-metadata-specs-tests/src/test/kotlin/com/squareup/kotlinpoet/metadata/specs/test/KotlinPoetMetadataSpecsTest.kt]: Constructor exception didn't get recognized?

File: kotlinpoet-metadata-specs/src/main/kotlin/com/squareup/kotlinpoet/metadata/specs/KotlinPoetMetadataSpecs.kt [hunk 3]
```diff
@@ -816,6 +826,19 @@ private inline fun <E> setOf(body: MutableSet<E>.() -> Unit): Set<E> {
...
 private val JVM_DEFAULT = JvmDefault::class.asClassName()
 private val JVM_STATIC = JvmStatic::class.asClassName()
+private fun createThrowsSpec(
+  exceptions: Set<TypeName>,
+  useSiteTarget: UseSiteTarget? = null
+): AnnotationSpec {
+  return AnnotationSpec.builder(Throws::class)
+      .addMember(
+          "exceptionClasses = %L",
+          exceptions.map { CodeBlock.of("%T::class", it) }
```
Comment (Egorand) [inline kotlinpoet-metadata-specs/src/main/kotlin/com/squareup/kotlinpoet/metadata/specs/KotlinPoetMetadataSpecs.kt]: nit: These won't get new lines automatically if the array is too long, this would require some manual formatting. Not important for this PR.

File: kotlinpoet-metadata-specs/src/main/kotlin/com/squareup/kotlinpoet/metadata/specs/ElementHandler.kt (hunk 1 above)
Comment (ZacSweers) [reply kotlinpoet-metadata-specs/src/main/kotlin/com/squareup/kotlinpoet/metadata/specs/ElementHandler.kt]: null allows us to lean on kotlin nullability rather than checking if the returned annotations are empty

File: kotlinpoet-metadata-specs-tests/src/test/kotlin/com/squareup/kotlinpoet/metadata/specs/test/KotlinPoetMetadataSpecsTest.kt (hunk 2 above)
Comment (ZacSweers) [reply kotlinpoet-metadata-specs-tests/src/test/kotlin/com/squareup/kotlinpoet/metadata/specs/test/KotlinPoetMetadataSpecsTest.kt]: oooo good catch! Will look

File: kotlinpoet-metadata-specs/src/main/kotlin/com/squareup/kotlinpoet/metadata/specs/ElementHandler.kt (hunk 1 above)
Comment (JakeWharton) [reply kotlinpoet-metadata-specs/src/main/kotlin/com/squareup/kotlinpoet/metadata/specs/ElementHandler.kt]: Collections should almost never be nullable unless there's some semantic meaning between non-null vs. null and empty vs. non-empty. For things like an allow/disallow list, null might mean allow all whereas empty list means allow nothing.

Since you don't need the ternary state, empty list makes more sense here. You already always have the problem of dealing with empty lists in collections. Adding null just complicates things.

Comment (ZacSweers) [reply kotlinpoet-metadata-specs/src/main/kotlin/com/squareup/kotlinpoet/metadata/specs/ElementHandler.kt]: fair enough, will change 👍

**Comment by ZacSweers** [comment]
Fixed missing constructor annotations and changed to use a non-nullable return type.

I also found a bug in `ParameterSpec` where primary constructors with no keyword emission would still emit parentheses even with no parameters. Technically legal in kotlin, but redundant

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/ParameterSpec.kt [hunk 4]
```diff
@@ -183,9 +183,13 @@ class ParameterSpec private constructor(
 internal fun List<ParameterSpec>.emit(
   codeWriter: CodeWriter,
   forceNewLines: Boolean = false,
+  forceParensOnEmpty: Boolean = true,
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/ParameterSpec.kt]: I'd really prefer these changes to be in a separate PR, looks they're not strictly related to this one. Can you please pull them out and rebase this PR on top? Should be an easy merge.

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeSpec.kt [hunk 5]
```diff
@@ -473,7 +473,7 @@ class TypeSpec private constructor(
 
     fun primaryConstructor(primaryConstructor: FunSpec?) = apply {
       check(kind == Kind.CLASS) {
-        "$kind can't have initializer blocks"
+        "$kind can't have a primary constructor"
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: Also feels like a separate PR, even if a very small one.

Comment (ZacSweers) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: https://github.com/square/kotlinpoet/pull/773

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/ParameterSpec.kt (hunk 4 above)
Comment (ZacSweers) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/ParameterSpec.kt]: https://github.com/square/kotlinpoet/pull/774

**Comment by Egorand** [comment]
Looks good! Wanna rebase against master before I merge?

**Comment by ZacSweers** [comment]
Rebased!
